src/charm.py

```diff
@@ -125,9 +156,146 @@ def _configure_nrf(self, event: EventBase) -> None:
             return
         needs_restart = self._generate_config_file()
         self._configure_workload(restart=needs_restart)
-        self._publish_nrf_info_for_all_requirers(NRF_URL)
+        self._publish_nrf_info_for_all_requirers()
         self.unit.status = ActiveStatus()
 
+    def _on_certificates_relation_created(self, event: EventBase) -> None:
+        """Generates Private key."""
+        if not self.unit.is_leader():
+            return
+        if not self._container.can_connect():
+            event.defer()
+            return
+        self._generate_private_key()
```

Comment:
This will regenerate a private key if the integration is removed and added back. It is probably not a problem, but it might be better to put the private key in a secret and use the rotation features (when they will work).

Reply:
Yes the private key would be removed when the relation is removed, same for the cert. We don't need the private key without the relation to a tls provider.